### PR TITLE
BUG: special: Fix numerical precision issue of log_ndtr

### LINKED ISSUE #15172: BUG: special: High relative error in `log_ndtr`
### Describe your issue.

The function `scipy.special.log_ndtr` (which computes the log of the CDF of the standard normal distribution) has an unacceptably high relative error in the interval [5, 6].  The following plot of relative errors for the interval [-35, 35] (generated by the script below, which uses `mpmath` to compute the true values), illustrates the problem:

log_ndtr_rel_error_1 9 0 dev0+1033 5bfbe97

We don't have well-defined criteria for how large relative error must be for it to be called a bug, but I think the behavior in the interval [5, 6], where the relative error can be as high as 5e-8, qualifies.

The current implementation uses the naive calculation `log(ndtr(x))` for x in [-20, 6].  This loses precision when x is positive, and the relative error grows rapidly as x increases.

The calculation can be reformulated using the functions `erfc` and `erfcx` to avoid this extreme loss of precision.  I will submit a pull request with a fix.

### Reproducing Code Example

```python
import numpy as np
import scipy
from scipy.special import log_ndtr
import mpmath
import matplotlib.pyplot as plt


def log_ndtr_mp(x):
    return mpmath.log(mpmath.ncdf(x))


mpmath.mp.dps = 400

intervals = [(-35, -1), (-1, 2.5), (2.5, 7.5), (7.5, 35)]

fig, axes = plt.subplots(2, len(intervals)//2, figsize=(9, 5),
                         constrained_layout=True)

for (a, b), axis in zip(intervals, axes.ravel()):
    x = np.linspace(a, b, 1250)
    ymp = np.array([float(log_ndtr_mp(t)) for t in x])
    y = log_ndtr(x)
    relerr = np.abs((y - ymp)/ymp)

    axis.plot(x, relerr)
    axis.grid()

fig.suptitle('Relative error of log_ntr(x), '
             f'SciPy version {scipy.__version__}\n ')
# plt.show()
plt.savefig(f'log_ndtr_rel_error_{scipy.__version__}.png')
```

### Error message

```shell
None.
```

### SciPy/NumPy/Python version information

1.9.0.dev0+1033.5bfbe97 1.21.2 sys.version_info(major=3, minor=9, micro=5, releaselevel='final', serial=0)

### PR DESCRIPTION
**Pull request by WarrenWeckesser**
Reimplement `log_ndtr` to maintain consistently high precision.

The implementation is explained in the comments above the function
`faddeeva_log_ndtr()` in the file `_faddeeva.cxx`.

The test `TestNdtriExp.test_extreme` was rewritten. It tests the
round-trip calculation `x2 = log_ndtr(ndtri_exp(x))` at two extreme
input values.  The change of one of the extreme input values is
explained in the long comment in the test.

The other extreme value is changed to `-np.finfo(float).tiny`.  This
is because it made no sense to test that round-trip with the huge value
`-np.finfo(float).min`, which is the same as `np.finfo(float).max`,
approx. 1.8e308.  `ndtri_exp` returns `nan` for inputs greater than 0,
so the round-trip calculation returned `nan`.  The test passed because
the comparison function `assert_func_equal()` was used with `nan_ok=True`,
so the `nan` result was ignored (and so the test was pointless).  My
guess is the intent was to test the value that is now there, namely
`-np.finfo(float).tiny`, which is approximately -2.225e-308.  Because
this test is just a numerical comparison of two input values with two
output values, the simpler function `assert_allclose` is now used to
compare the results.

After updating `log_ndtr`, the stats test `TestExponNorm.test_std_pdf`
in `test_distributions.py` failed for two sets of parameter inputs,
(x, K) = (1, 0.01) and (x, K) = (-1, 0.01).  The calculation of
`exponnorm._logpdf` calls `log_ndtr(x - 1/K)` (via `_norm_logcdf`, which
is just a one-line wrapper of `log_ndtr`), and for these two sets of
parameters, the calculation of `log_ndtr(x - 1/K)` is *more* accurate
with this change than before.  So the increased error observed in the
test is the result of some other numerical effect.

Also, an 'Examples' section was added to the docstring of `log_ndtr`.

Closes gh-15172

### CODE REVIEW

**Comment by WarrenWeckesser** [comment]
Here's a copy of the plot of the relative errors of `log_ndtr(x)` that I included in gh-15172:

log_ndtr_rel_error_1 9 0 dev0+1033 5bfbe97

Here's the result of the same plotting script when run with this PR.  Note the changes in the vertical scales:

log_ndtr_rel_error_1 9 0 dev0+1043 71c8003

By the way, the formulation used in `faddeeva_log_ndtr` is the same as that in the Julia code in `Distributions.jl`:

https://github.com/JuliaStats/Distributions.jl/blob/d308d9bf5903b643c4dde13011c97ac38f9c0cbf/src/univariate/continuous/normal.jl#L147-L154

File: scipy/special/tests/test_ndtri_exp.py [hunk 1]
```diff
@@ -60,13 +60,32 @@ def test_in_interval(self, interval, expected_rtol, uniform_random_points):
...
 
     def test_extreme(self):
-        assert_func_equal(
-            log_ndtr_ndtri_exp,
-            lambda y: y,
-            [-np.finfo(float).max, -np.finfo(float).min],
-            rtol=1e-12,
-            nan_ok=True
-        )
+        # bigneg is not quite the largest negative double precision value.
```
Comment (steppi) [inline scipy/special/tests/test_ndtri_exp.py:63]: This is a really good consideration. I wrote this test kind of as an after-thought and didn't pay sufficient attention that everything actually made sense. Sigh, that `nan_ok` is practically unforgivable. Thanks for the in-depth explanation.

File: scipy/special/tests/test_ndtri_exp.py [hunk 2]
```diff
@@ -60,13 +60,32 @@ def test_in_interval(self, interval, expected_rtol, uniform_random_points):
...
+        # finite value that can be represented with 64 bit float point.  This
+        # means tiny changes in how the computation proceeds can result in the
+        # return value being -inf.  (E.g. changing the constant representation
+        # of 1/sqrt(2) from 0.7071067811865475--which is the value returned by
+        # 1/np.sqrt(2)--to 0.7071067811865476--which is the most accurate 64
+        # bit floating point representation of 1/sqrt(2)--results in the
+        # round-trip that starts with np.finfo(float).min returning -inf.  So
+        # we'll move the bigneg value a few ULPs towards 0 to avoid this
+        # sensitivity.
+        # Use the reduce method to apply nextafter four times.
```
Comment (steppi) [inline scipy/special/tests/test_ndtri_exp.py:82]: This is a really good consideration. I wrote this test kind of as an after-thought and didn't pay sufficient attention that everything actually made sense. Sigh, that `nan_ok` is practically unforgivable. Thanks for the in-depth explanation.

**Comment by WarrenWeckesser** [comment]
Merged main; lots of test failures, but they all look like the common CI issues that have been occurring lately (e.g. https://github.com/scipy/scipy/issues/15854) and are not related to this PR.

File: scipy/special/tests/test_ndtri_exp.py (hunk 2 above)
Comment (person142) [reply scipy/special/tests/test_ndtri_exp.py:82]: I wouldn't worry to much about it - that `nan_ok` has bitten us many a time. I once had ambitions to turn it off everywhere and then remove that flag, but that turned out to be more complex than hoped. A change that flips it to being default off would be super smooth though!

**Review by person142 — APPROVED** [review]

**Comment by person142** [comment]
Thanks @WarrenWeckesser! Gonna take your word re CI failures; am not keeping up with that fickle beast these days.

**Pull request merged**